### DISCUSSION TITLE
volatile type of variable

### DISCUSSION THREAD

**Discussion by guilleruiz09** [TheAlbertDev/MASB#84] [Q&A] [answered]
Hello,
I do not totally understand the ```volatile``` type of variables used in arduino. When reading the explanation I see that it makes the compiler load the variable from the RAM, instead of from the storage register. This is done to avoid using a value of a variable which could not be the current one, as it could be in the register due to it having been called by the loop. However, if the loop has stored this new value in the register I understand that we would want to access it instead of the RAM, but it is just opposite of what this ```volatile``` does. I am sure that some part of my reasoning is wrong but I do not know where.
Thanks in advance to anyone wanting to help.

**Comment by alvaroleva** [comment, marked as answer]
Hi Guillermo,
I hope my answer clears up the concept.  As you said, variables defined with the `volatile` qualifier are not loaded from a storage register, but from RAM memory.  By default, if you don't declare it as `volatile`, the microprocessor accesses the value of the variable by storing it in a register because it is quicker in comparison to accessing it from RAM memory (a way of optimizing time). However, this **can lead to problems** when the variable is at the same time being updated outside of the main program flow. This occurs, for instance, **when you apply an interrupt and your ISR modifies the value of a variable that is already being used in another function inside the program.**  In that case, if you don't declare the variable as `volatile` the microprocessor may experience a delay while trying to write the ISR updated value from the register to the memory. Consequently, it can lead to the value in the register being OUTDATED with the value in memory. 
Summing up, every time that you want to use a variable that can be edited simultaneously from two functions (such as an `estadoLED` that is used in the void loop and edited with the button interrupt ISR) you must declare it with the `volatile` qualifier. In this way, you will ensure that the microcontroller is accessing the value directly from the RAM memory and, although being worse in terms of performance, it ensures using the last updated value. 

Here's a [link](https://www.tutorialspoint.com/difference-between-register-and-memory) where I found the differences between register and memory storage. 

I hope it helps,
Álvaro.
